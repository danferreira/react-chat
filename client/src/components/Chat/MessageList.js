import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import SmartScroll from '../SmartScroll/SmartScroll';
import Message from './Message';

const List = styled.div`
    background: #353838;
    order: 2;
    height: 100vh;
    position: relative;    
    display: flex;
    justify-content: center;
    align-items: center;
`

const propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }))
}

const defaultProps = {
    messages: []
}

class MessageList extends Component {

    handleScrollTop = () => {
        const { isLoadingMoreItems, onLoadMoreItems } = this.props;

        if (onLoadMoreItems && !isLoadingMoreItems) {
            onLoadMoreItems();
        }
    }

    render() {
        const { messages, isLoadingMoreItems } = this.props;

        return (
            <List>                
                <SmartScroll 
                    onScrollTop={this.handleScrollTop}
                    isLoadingMoreItems={isLoadingMoreItems}>
                    {messages.map((m) =>
                        <Message
                            message={m}
                            key={m.id} />
                    )}
                </SmartScroll>
            </List>
        )
    }
}

MessageList.propTypes = propTypes;
MessageList.defaultProps = defaultProps;

export default MessageList;