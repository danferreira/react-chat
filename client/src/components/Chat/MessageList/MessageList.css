.message-list {
    background: #353838;
    order: 2;
    height: 100vh;
    position: relative;    
    display: flex;
    justify-content: center;
    align-items: center;
}