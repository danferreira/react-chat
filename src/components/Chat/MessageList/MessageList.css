.message-list {
    background: #353838;
    order: 2;
    height: 100vh;
    position: relative;
}