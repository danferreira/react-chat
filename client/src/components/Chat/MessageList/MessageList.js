import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SmartScroll from '../../SmartScroll/SmartScroll';
import Message from './Message/Message';
import './MessageList.css';
import Spinner from '../../Spinner/Spinner';

const propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }))
}

const defaultProps = {
    messages: []
}

class MessageList extends Component {

    handleScrollTop = () => {
        const { hasMoreItems, isLoadingMoreItems, onLoadMoreItems } = this.props;

        if (hasMoreItems && !isLoadingMoreItems) {
            onLoadMoreItems();
        }
    }

    render() {
        const { messages, isLoadingMoreItems } = this.props;

        return (
            <div className="message-list">                
                <SmartScroll 
                    onScrollTop={this.handleScrollTop}
                    isLoadingMoreItems={isLoadingMoreItems}>
                    {messages.map((m) =>
                        <Message
                            message={m}
                            key={m.id} />
                    )}
                </SmartScroll>
            </div>
        )
    }
}

MessageList.propTypes = propTypes;
MessageList.defaultProps = defaultProps;

export default MessageList;