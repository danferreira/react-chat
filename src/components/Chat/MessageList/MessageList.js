import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Loader from '../../Loader/Loader';
import SmartScroll from '../../SmartScroll/SmartScroll';
import Message from './Message/Message';
import './MessageList.css';

const propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        content: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired
    }))
}

const defaultProps = {
    messages: []
}

class MessageList extends Component {

    render() {
        const { messages } = this.props;

        return (
            <div className="message-list">
                <Loader isLoading={messages.length === 0}>
                    <SmartScroll>
                        {messages.map((m) =>
                            <Message
                                message={m}
                                key={m.id} />
                        )}
                    </SmartScroll>
                </Loader>
            </div>
        )
    }
}

MessageList.propTypes = propTypes;
MessageList.defaultProps = defaultProps;

export default MessageList;