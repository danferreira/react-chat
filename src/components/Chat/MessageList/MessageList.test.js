import React from 'react';
import { shallow } from 'enzyme';

import MessageList from './MessageList';

describe('<MessageList />', () => {

    it('should render an empty list', () => {
        var wrapper = shallow(<MessageList />);
        expect(wrapper).toMatchSnapshot();
    })

    it('should render a message list', () => {
        var messages = [{
            id: "1",
            content: "hello",
            type: "out"
        },
        {
            id: "2",
            content: "hi. How are you?",
            type: "in"
        }];

        var wrapper = shallow(<MessageList messages={messages}/>);
        expect(wrapper).toMatchSnapshot();
    })
});