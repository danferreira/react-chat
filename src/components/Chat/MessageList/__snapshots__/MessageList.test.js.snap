// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList /> should render a message list 1`] = `
<div
  className="message-list"
>
  <SmartScroll>
    <Message
      key="1"
      message={
        Object {
          "content": "hello",
          "id": "1",
          "type": "out",
        }
      }
    />
    <Message
      key="2"
      message={
        Object {
          "content": "hi. How are you?",
          "id": "2",
          "type": "in",
        }
      }
    />
  </SmartScroll>
</div>
`;

exports[`<MessageList /> should render an empty list 1`] = `
<div
  className="message-list"
>
  <SmartScroll />
</div>
`;
